libs/beacon/src/beacon_setup_service.cpp

```diff
@@ -128,238 +169,771 @@ BeaconSetupService::State BeaconSetupService::OnIdle()
-  std::unordered_set<Address> peers(v_peers.begin(), v_peers.end());
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_DIRECT_CONNECTIONS));
 
-  bool     all_connected = true;
-  uint16_t connected     = 0;
+  auto                              v_peers = endpoint_.GetDirectlyConnectedPeers();
+  std::unordered_set<MuddleAddress> peers(v_peers.begin(), v_peers.end());
+
+  std::set<MuddleAddress> connected;
 
   for (auto &m : beacon_->aeon.members)
   {
     // Skipping own address
     if (m == identity_)
     {
+      connected.insert(m.identifier());
       continue;
     }
 
     // Checking if other peers are there
     if (peers.find(m.identifier()) == peers.end())
     {
-      all_connected = false;
-
       // TODO(tfr): Request muddle to connect.
     }
     else
     {
-      connected++;
+      connected.insert(m.identifier());
     }
   }
 
-  if (all_connected)
+  if (connected.size() == beacon_->aeon.members.size())
   {
-    FETCH_LOG_INFO(LOGGING_NAME, "All peers connected. Proceeding.");
-    return State::BROADCAST_ID;
+    connections_ = connected;
+    fetch::serializers::MsgPackSerializer serializer;
+    serializer << connected;
+    pre_dkg_rbc_.Broadcast(serializer.data());
+    return State::WAIT_FOR_READY_CONNECTIONS;
   }
 
   state_machine_->Delay(std::chrono::milliseconds(200));
-  FETCH_LOG_INFO(LOGGING_NAME,
-                 "Waiting for all peers to join before starting setup. Connected: ", connected,
-                 " expect: ", beacon_->aeon.members.size() - 1);
+  FETCH_LOG_INFO(LOGGING_NAME, "Waiting for all peers to join before starting setup. Connected: ",
+                 connected.size(), " expect: ", beacon_->aeon.members.size() - 1);
 
   return State::WAIT_FOR_DIRECT_CONNECTIONS;
 }
 
-BeaconSetupService::State BeaconSetupService::OnBroadcastID()
+BeaconSetupService::State BeaconSetupService::OnWaitForReadyConnections()
 {
   std::lock_guard<std::mutex> lock(mutex_);
+  dkg_state_gauge_->set(static_cast<uint8_t>(State::WAIT_FOR_READY_CONNECTIONS));
 
-  // Creating identity
-  CabinetMemberDetails member;
-  member.identity = beacon_->manager.identity();
-  member.id       = beacon_->manager.id();
-  member_details_queue_.push_back(member);
+  // Try broadcasting connections again in case they were discarded by RBCs who's cabinet was reset
+  // late
+  fetch::serializers::MsgPackSerializer serializer;
+  serializer << connections_;
+  pre_dkg_rbc_.Broadcast(serializer.data());
 
-  Serializer msgser;
-  msgser << member;
+  if (ready_connections_.size() >= beacon_->aeon.members.size() - 1)
+  {
+    for (auto &m : beacon_->aeon.members)
+    {
+      if (m == identity_)
+      {
+        continue;
+      }
+      if (ready_connections_.find(m.identifier()) != ready_connections_.end())
+      {
+        assert(ready_connections_.at(m.identifier()) == connections_);
+        // TODO(jmw): Strategy if connections for members are different
```

Comment:
Yeah, we should think a bit about this tomorrow perhaps - are the conditions for continuing that threshold number of members form a complete subgraph?